src/sp/protocol/mqtt/nmq_mqtt.c

```diff
@@ -508,15 +548,50 @@ nano_pipe_init(void *arg, nni_pipe *pipe, void *s)
 	nni_aio_init(&p->aio_timer, nano_pipe_timer_cb, p);
 	nni_aio_init(&p->aio_recv, nano_pipe_recv_cb, p);
 
-	p->reason_code             = 0x00;
-	p->id                      = nni_pipe_id(pipe);
+	p->conn_param = nni_pipe_get_conn_param(pipe);
+	p->id = nni_pipe_id(pipe);
 	p->pipe                    = pipe;
+	clientid      = (char *) conn_param_get_clientid(p->conn_param);
+	// restore session according to clientid
+	if (clientid && (p->conn_param->clean_start == 0)) {
+		nano_pipe *old;
+		clientid_key = DJBHashn(clientid, strlen(clientid));
+		old = nni_id_get(&sock->cached_sessions, clientid_key);
+		if (old != NULL) {
+			// replace nano_qos_db and pid with old one.
+			p->pipe->packet_id = old->pipe->packet_id;
+			nni_id_map_fini(p->pipe->nano_qos_db);
+			nng_free(p->pipe->nano_qos_db, sizeof(struct nni_id_map));
+			p->pipe->nano_qos_db = old->nano_qos_db;
+			nni_pipe_id_swap(pipe->p_id, old->pipe->p_id);
+			p->id = old->pipe->p_id;
+			pipe->p_id = old->pipe->p_id;
+			old->pipe->cache = false;
+			nni_id_remove(&sock->cached_sessions, clientid_key);
+			// close old one (bool to prevent disconnect_ev)
+			// check if pointer is different later
+			nni_pipe_close(old->pipe);
+		}
+	} else {
+		// clean previous session
+		nano_pipe *old;
+		clientid_key = DJBHashn(clientid, strlen(clientid));
```

Comment:
The client id should be checked

Reply:
nope no null clientid would present. because transport assgined one